Allow non-hash based blocknum argument in CeloAPI backend

ezdac opened the pull request:
Fixes #136 

- removes the LRU cache in the `CeloAPIBackend.GetExchangeRates` call because of premature optmisation
- adds a function `rpc.BlockNumberOrHashEnsureHashOnly`:
  - converts number and `latest, pending, earliest, ...`  based block-number arguments to their hash based on the blockchain state
  - has extra argument flags that force the returned hash to be the block's parent hash
- changes the call signature of `ethapi.CeloBackend` methods to accept `rpc.BlockNumberOrHash` instead of `common.Hash` for specifying blocks to query
- uses `rpc.BlockNumberOrHashEnsureHashOnly` to convert `pending` RPC arguments in `eth_estimateGas` to the pending block's parent hash, so that state-db queries don't fail (#136)

Unfortunately the fix for #136 is not easily tested in a unit-test, since the `ethapi/api_test.go`'s mock backend does not fully implement querying state of the `pending` block. We should re-evaluate in the future, if it is worth to change the mock implementation here.

ezdac committed:
Allow non-hash based blocknum argument in CeloAPI backend

palango commented on internal/celoapi/backend.go:
```diff
@@ -37,55 +37,60 @@ type CeloAPIBackend struct {
 	}
-	b.exchangeRatesCache.Add(atBlock, er)
+	if isHash {
+		b.exchangeRatesCache.Add(blockHash, er)
```
I wonder if we should add some tests for the caching here.

ezdac replied:
I removed the whole LRU cache 514de826181d. The usefulness should be decided at a later point, and in general this
should probably be implemented on a lower level anyways.

palango commented on internal/ethapi/api.go:
```diff
@@ -1722,10 +1725,11 @@ func AccessList(ctx context.Context, b CeloBackend, blockNrOrHash rpc.BlockNumbe
-			// queried block, but want to use the exchange rates before (at the beginning of)
-			// the queried block
-			exchangeRates, err = b.GetExchangeRates(ctx, header.ParentHash)
+			feeBlockNum, err := rpc.BlockNumberOrHashEnsureHashOnly(ctx, b, blockNrOrHash, true, true)
```
It would be good to carry over the comment, especially in how it related to the last two parameters here (and in the call above).

ezdac replied:
Done in e11a9caa2b0f

carterqw2 commented on internal/ethapi/api.go:
```diff
@@ -1286,12 +1286,15 @@ func DoEstimateGas(ctx context.Context, b CeloBackend, args TransactionArgs, blo
+		if err != nil {
+			return 0, err
+		}
+		exchangeRates, err = b.GetExchangeRates(ctx, feeBlockNum)
```
If I read the code correctly, after you changed the implementation of `GetExchangeRate`, at this point it's safe to just pass `blockNrOrHash` here. In this case `StateAndHeaderByNumberOrHash` should correctly fetch the state for the `latest` block instead of `pending` as `blockNrOrHash` would be a number in this case.

ezdac replied:
Yes you are right that it is safe to just pass `blockNrOrHash`. If the pending state can't be retrieved it will use latest as a fallback.
However there is the tiny difference, that since now the pending state **is** retrievable, we use the current pending state during block building for gas estimation and the *newly updated exchange rates* that already applied during the block building. This is inconsistent with what we usually define as the exchange rates for a block, namely "freezing" the state to the end-state of the previously sealed block.

At this place this might not be too critical, since this function is estimating the gas-limit and thus an inconsistent exchange rate only has an effect on the maximum allowance in the gas estimation and potential execution failure due to insufficient funds.

ezdac replied:
Done in e11a9caa2b0f

I decided to not go with the `pending` block distinction and use the passed in block itself for all queries. `Pending` usually means "the state during block building" so using the most recent exchange rates seemed more appropriate.

carterqw2 commented on internal/ethapi/api.go:
```diff
@@ -1722,10 +1725,11 @@ func AccessList(ctx context.Context, b CeloBackend, blockNrOrHash rpc.BlockNumbe
+			if err != nil {
+				return nil, 0, nil, fmt.Errorf("query hash of provided block argument: %w", err)
+			}
+			exchangeRates, err = b.GetExchangeRates(ctx, feeBlockNum)
```
Why can't we use `rpc.BlockNumberOrHashWithHash(header.ParentHash, false)`? Maybe we can even [drop](https://github.com/celo-org/op-geth/compare/ezdac/issue-136...celo-org:op-geth:carterqw2/issue-136?expand=1) the function altogether.

ezdac replied:
True, this is simpler.
I was not too happy with this functions call signature - I will drop the function and implement the `pending` logic directly within the `DoEstimate` function (see above / other comment).

ezdac replied:
Done in e11a9caa2b0f

ezdac committed:
fixup! Allow non-hash based blocknum argument in CeloAPI backend

carterqw2 reviewed (approved):
Nice!

ezdac committed:
Remove LRU cache from GetExchangeRates

Lets keep it simple and not do any premature optimization.
If we implement this it might be better to do it on a lower level
anyways.
